From 5a3a4cf3015daa531c5cc0fa59890c221b217b52 Mon Sep 17 00:00:00 2001
From: Matthew Dillon <dillon@apollo.backplane.com>
Date: Mon, 25 Jul 2016 11:31:04 -0700
Subject: [PATCH 066/100] kernel - Fix mountctl() / unmount race

* kern_mountctl() now properly checks to see if an unmount is in-progress
  and returns an error, fixing a later panic.
---
 sys/kern/vfs_syscalls.c | 12 +++++++++++-
 1 file changed, 11 insertions(+), 1 deletion(-)

diff --git a/sys/kern/vfs_syscalls.c b/sys/kern/vfs_syscalls.c
index b0430e8..41374f8 100644
--- a/sys/kern/vfs_syscalls.c
+++ b/sys/kern/vfs_syscalls.c
@@ -1106,6 +1106,7 @@ kern_mountctl(const char *path, int op, struct file *fp,
 	struct vnode *vp;
 	struct nlookupdata nd;
 	struct nchandle nch;
+	struct mount *mp;
 	int error;
 
 	*res = 0;
@@ -1135,6 +1136,8 @@ kern_mountctl(const char *path, int op, struct file *fp,
 	nlookup_done(&nd);
 	vn_unlock(vp);
 
+	mp = nch.mount;
+
 	/*
 	 * Must be the root of the filesystem
 	 */
@@ -1143,7 +1146,14 @@ kern_mountctl(const char *path, int op, struct file *fp,
 		vrele(vp);
 		return (EINVAL);
 	}
-	error = vop_mountctl(nch.mount->mnt_vn_use_ops, vp, op, fp, ctl, ctllen,
+	if (mp == NULL || mp->mnt_kern_flag & MNTK_UNMOUNT) {
+		kprintf("kern_mountctl: Warning, \"%s\" racing unmount\n",
+			path);
+		cache_drop(&nch);
+		vrele(vp);
+		return (EINVAL);
+	}
+	error = vop_mountctl(mp->mnt_vn_use_ops, vp, op, fp, ctl, ctllen,
 			     buf, buflen, res);
 	vrele(vp);
 	cache_drop(&nch);
-- 
2.7.2

